activation/nipost.go

```diff
@@ -377,64 +356,130 @@ func (nb *NIPostBuilder) submitPoetChallenge(
 
 	round, err := client.Submit(submitCtx, deadline, prefix, challenge, signature, nodeID)
 	if err != nil {
-		return &PoetSvcUnstableError{msg: "failed to submit challenge to poet service", source: err}
+		return nipost.PoETRegistration{},
+			&PoetSvcUnstableError{msg: "failed to submit challenge to poet service", source: err}
 	}
 	logger.Info("challenge submitted to poet proving service", zap.String("round", round.ID))
-	return nipost.AddPoetRegistration(nb.localDB, nodeID, nipost.PoETRegistration{
+
+	registration := nipost.PoETRegistration{
 		ChallengeHash: types.Hash32(challenge),
 		Address:       client.Address(),
 		RoundID:       round.ID,
 		RoundEnd:      round.End,
-	})
+	}
+
+	if err := nipost.AddPoetRegistration(nb.localDB, nodeID, registration); err != nil {
+		return nipost.PoETRegistration{}, err
+	}
+
+	return registration, err
 }
 
-// Submit the challenge to all registered PoETs.
+// submitPoetChallenges submit the challenge to registered PoETs
+// if some registrations are missing and PoET round didn't start.
 func (nb *NIPostBuilder) submitPoetChallenges(
 	ctx context.Context,
 	signer *signing.EdSigner,
-	deadline time.Time,
+	poetProofDeadline time.Time,
+	curPoetRoundStartDeadline time.Time,
 	challenge []byte,
-) error {
+) ([]nipost.PoETRegistration, error) {
+	// check if some registrations missing or were removed
+	nodeID := signer.NodeID()
+	registrations, err := nipost.PoetRegistrations(nb.localDB, nodeID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get poet registrations from db: %w", err)
+	}
+
+	registrationsMap := make(map[string]nipost.PoETRegistration)
+	for _, reg := range registrations {
+		registrationsMap[reg.Address] = reg
+	}
+
+	existingRegistrationsMap := make(map[string]nipost.PoETRegistration)
+	var missingRegistrations []PoetService
+	for addr, poet := range nb.poetProvers {
+		if val, ok := registrationsMap[addr]; ok {
+			existingRegistrationsMap[addr] = val
+		} else {
+			missingRegistrations = append(missingRegistrations, poet)
+		}
+	}
+
+	existingRegistrations := maps.Values(existingRegistrationsMap)
+	if len(missingRegistrations) == 0 {
+		return existingRegistrations, nil
+	}
+
+	now := time.Now()
+
+	if curPoetRoundStartDeadline.Before(now) {
+		if len(existingRegistrations) == 0 {
+			if len(registrations) == 0 {
+				// no existing registration at all, drop current registration challenge
+				err = ErrATXChallengeExpired
+			} else {
+				// no existing registration for given poets set
+				err = ErrNoRegistrationForGivenPoetFound
+				nb.logger.Warn("revert poet configuration to previous is needed immediately",
+					zap.Error(err), log.ZShortStringer("smesherID", nodeID))
+			}
+			return nil, fmt.Errorf(
+				"%w: poet round has already started at %s (now: %s)",
+				err,
+				curPoetRoundStartDeadline,
+				now,
+			)
```

Comment:
This will cause an endless loop if `existingRegistrations == 0` and `len(registrations) != 0`. Instead the node should in my opinion exit here with a `log.Fatal` or similar.

Reply:
Do you mean, that submit challenge will endless try to submit challenge in that case? 
The main point here return `ErrNoRegistrationForGivenPoetFound` in case if it is total mismatch between registrations, which we already have in db, and poets, which we have in config. So it is the case, when user has registrations, but wouldn't be able to fetch the proof. 

I am not sure, if log.Fatal is needed here. @poszu WDYT?

Reply:
> Do you mean, that submit challenge will endless try to submit challenge in that case?

Yes.

> I am not sure, if log.Fatal is needed here.

The node is in a non functional state - it has a config that doesn't fit to its local state. Just logging a warning is in my opinion not sufficient, since for me this is clearly a misconfigured node that should exit to make sure the operator / user fixes the issue and doesn't ignore the warning that we print.

Reply:
I think @fasmat is right here. Ending up with some registrations in the DB, but not using any of them (because of the config) is most likely a misconfiguration and it needs a user action to be fixed.